### PR TITLE
[litmus] Print instruction values with offset argument

### PR DESCRIPTION
**Pull request by maranget**
This is a followup to PR #547. Litmus and herd should now print instruction values identically.

### CODE REVIEW

**Review by relokin — COMMENTED** [review]
A few minor comments, feel free to ignore, otherwise this PR looks good to me.

File: lib/CBase.ml [hunk 1]
```diff
@@ -344,6 +344,7 @@ include Pseudo.Make
         fun i -> get_rec 0 i
 
 
+      let size_of_ins _ = 4
```
[the remark this answers is not in the document]
Comment (maranget) [reply lib/CBase.ml:347]: Ok I was naive, the code loader of course fails. You are right: value 1 is a more plausible choice.

File: lib/X86Base.ml [hunk 2]
```diff
@@ -588,6 +588,8 @@ include Pseudo.Make
             -> get_naccs_eff e
         | I_MOVSD -> 2
 
+      let size_of_ins _ = 4
```
[the remark this answers is not in the document]
Comment (maranget) [reply lib/X86Base.ml:591]: Indeed, this is not correct. We may fix this, some day. Notice, if I am not wrong that exact instruction sizes at the litmus level are strictly useful (_i.e._ necessary) for self modifying code (to compute code size) and  useful for label initial values (such as `0:X0=P0:L0`). In the non-self mode, label initial values could probably be implemented by global labels and not by offsets. Some comment that  point at the meaningless value for instruction size should probably be added here and in `lib/X86_64Base.ml`.

File: herd/loader.ml [hunk 3]
```diff
@@ -41,74 +42,98 @@ struct
...
-      | A.Symbolic _
-      | A.Macro (_,_) -> assert false
-      end
+  let rec preload_labels proc addr mem = function
+    | [] -> mem
+    | ins::code ->
+       begin
+         match ins with
+         | A.Nop ->
+            preload_labels proc addr mem code
```
[the remark this answers is not in the document]
Comment (maranget) [reply herd/loader.ml:51]: Those non-instruction "Nop"'' result from blank cells in concrete syntax (see example below) and are elided later. Hence their "zero-size".
```
AArch64 A
{
}
  P0       |  P1        ;
 MOV W0,#0 |            ;   (* NOP on P1 *)
           | MOV W1,#1  ;  (* NOP on P0 *)
```

File: herd/loader.ml [hunk 4]
```diff
@@ -41,74 +42,98 @@ struct
...
-        let next_mem, next_l2p =
-          if Label.Map.mem lbl mem then
-            Warn.user_error
-              "Label %s occurs more that once" lbl ;
-          Label.Map.add lbl addr mem , Label.Map.add lbl proc l2p in
-          preload_labels proc addr next_mem next_l2p (ins::code)
-      | A.Symbolic _
-      | A.Macro (_,_) -> assert false
-      end
+  let rec preload_labels proc addr mem = function
```
[the remark this answers is not in the document]
Comment (maranget) [reply herd/loader.ml]: Nice catch. The last commit below (9159c72f5c8ef94eb3c6035e32b437328ead9c74) improve code sharing by following your remark.

[the remark this answers is not in the document]
Comment (relokin) [reply herd/loader.ml]: Thanks! This was a bit tricker than I thought initially.

**Comment by relokin** [comment]
Thanks for all the changes Luc! This PR looks good to me!

**Comment by maranget** [comment]
I think this PR is ready. While testing it I found that the combination of `-variant kvm` and `-variant self` is not completely functional. Namely, some of the code adresses are computed in the source, static, code and not in the code copies, which they should. As a consequence, a final condition such as `fault(P0:L0,...)` is never satisfied. See issue #593.
```
AArch64 T
{
int x;
[PTE(x)]=(valid:0)
}

**Comment by maranget** [comment]
Merged and closed. We leave the fixing issue #593 to another PR.

**Pull request merged**